src/solvers/petsc_dm_wrapper.C

```diff
@@ -0,0 +1,388 @@
+// The libMesh Finite Element Library.
+// Copyright (C) 2002-2018 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+#include "libmesh/libmesh_common.h"
+
+#ifdef LIBMESH_HAVE_PETSC
+
+// PETSc includes
+#include <petscsf.h>
+
+#include "libmesh/petsc_dm_wrapper.h"
+
+#include "libmesh/system.h"
+#include "libmesh/mesh_base.h"
+#include "libmesh/dof_map.h"
+#include "libmesh/elem.h"
+
+namespace libMesh
+{
+
+PetscDMWrapper::~PetscDMWrapper()
+{
+  this->clear();
+}
+
+void PetscDMWrapper::clear()
+{
+  // This will also Destroy the attached PetscSection and PetscSF as well
+  // Destroy doesn't free the memory, but just resets points internally
+  // in the struct, so we'd still need to wipe out the memory on our side
+  for( auto dm_it = _dms.begin(); dm_it < _dms.end(); ++dm_it )
+    DMDestroy( dm_it->get() );
+
+  _dms.clear();
+  _sections.clear();
+  _star_forests.clear();
+}
+
+void PetscDMWrapper::init_and_attach_petscdm(const System & system, SNES & snes)
+{
+  START_LOG ("init_and_attach_petscdm", "PetscDMWrapper");
+
+  PetscErrorCode ierr;
+
+  // Eventually, we'll traverse the mesh hierarchy and cache
+  // for each grid level, but for now, we're just using the
+  // finest level
+  unsigned int n_levels = 1;
+  this->init_dm_data(n_levels);
+
+  for(unsigned int level = 0; level < n_levels; level++)
+    {
+      DM & dm = this->get_dm(level);
+      PetscSection & section = this->get_section(level);
+      PetscSF & star_forest = this->get_star_forest(level);
+
+      ierr = DMShellCreate(system.comm().get(), &dm);
```

Comment:
I was wondering why we do not implement a native libmesh DM? The native DM will give us more flexibility on the setup of the solver and the preconditioner.  In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future. 

We could clean up the old DM stuff, and combine with this PETScDMWrapper. It should be easy.

I miss something here?

Reply:
> I was wondering why we do not implement a native libmesh DM instead of shell dm?

Can you elaborate on what you mean by "native" DM? Do you mean to essentially do a full subclassing of DM?

> The native DM will give us more flexibility on the setup of the solver and the preconditioner. In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future.

Can you elaborate on what flexibility and customization is gained vs. the shell? As far as I can tell, we will get the full solver flexibility (once the part 2 comes with GMG). What else would be enabled that not's enabled by the shell?

Reply:
The other reason I like the shell interface is it minimizes the PETSc API we have to touch since PETSc has no intent so support backward compatibility in API so the more API we touch, the larger maintenance headache it becomes.

Reply:
> Can you elaborate on what you mean by "native" DM? Do you mean to essentially do a full subclassing of DM?

I mean something like  https://github.com/petsc/petsc/blob/65b40d628b4e55462e59718b6e7e25b2e90f85aa/include/petsc/private/dmpleximpl.h#L175

We may have something similar: DM_Libmesh.

> 
> The native DM will give us more flexibility on the setup of the solver and the preconditioner. In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future.
> Can you elaborate on what flexibility and customization is gained vs. the shell? As far as I can tell, we will get the full solver flexibility (once the part 2 comes with GMG). What else would be enabled that not's enabled by the shell?

Personally I like a native DM more than the shell DM because I could add some internal data (or other crazy things definitely at the user-side) in "typedef struct" that may be useful for implementing the solver interfaces. For example, to construct a geometric DDM, we may need to implement the interface: dm->ops->createdomaindecomposition. If we have our own internal data (such mesh info or system) inside of DM, it is straightforward to feedback the indices to PETSc, otherwise, we have to figure out a way (maybe PetscObjectCompose()) to bring in mesh or system info because the DDM interface does not have AppCtx.

Again, I could be wrong. All these stuffs are based on my experiences. I implemented GMG (a pure petsc-based code) for my  PhD thesis using the native DM.

Reply:
OK, so it looks like the existing DM bits were moving towards a "native" DM.

> Personally I like a native DM more than the shell DM because I could add some internal data (or other crazy things definitely at the user-side) in "typedef struct" that may be useful for implementing the solver interfaces. For example, to construct a geometric DDM, we may need to implement the interface: dm->ops->createdomaindecomposition. If we have our own internal data (such mesh info or system) inside of DM, it is straightforward to feedback the indices to PETSc, otherwise, we have to figure out a way (maybe PetscObjectCompose()) to bring in mesh or system info because the DDM interface does not have AppCtx.

Shell provides access to `createdomaindecomposition` and takes a context (step 2 of this process uses that context hook) so we can feed `EquationSystem` and therefore everything that would be needed. http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/DM/DMShellSetContext.html

That said, I'm not opposed to a native DM, but what's there looks rather cumbersome. Perhaps we could talk off thread about the DM API and what you're looking for. But in the near term, I'd prefer to stick to a shell so we can keep this moving.